ADBDEV-4950: Backport a patch for partition locking

bandetto opened the pull request:
Backport a patch for partition locking

A backport (b5c02fa) was merged in upstream, which prevented locks from being
released on partition children or their indexes before the end of transaction,
only for planner. We merged this patch with minor changes (2763ff1) for our
fork.

Then it’s reverted (d40e482) in upstream, but in our fork it’s not. The reason
for revert is that new changes break some test for indexes.

Last month, a different patch (4ac7227) is merged, which supposedly fixes the
parent issue for the planner and ORCA as well. It also does not change the
behavior of locks on indexes, like the first one did. It avoids releasing locks
for AO tables, and hides this ability behind a GUC, because we can exceed the
maximum amount of locks in a single transaction in cases with large partition
hierarchy.

Missing echo of DROP command was added to tests' output of the target patch.

2763ff1, 5894018 and b1373f8 were reverted and re-applied to fix conflicts.

This is a backport of 4ac7227.

maksm90 committed:
On DDL operation keep the lock on partitions until end of transaction (#295)

Commit fa287d0 had introduced early lock releasing on partitions within
operations with partitioned parent table. This was done to save memory
in lock manager I suppose. But such behavior incurs some issues, e.g.,
possible segfault when we perform drop of partitioned table and
concurrent `pg_total_relation_size()` call on some AO partition.

Obviously, we have to restore full locking for inheritance hierarchy as
it's done in vanilla PostgreSQL. To eliminate exhaustion of lock memory
we need to increase max_locks_per_transactions setting. Memory is cheap,
and if we're working with tens of thousands of partitions, we can afford
reserving some memory for the lock manager.

Partially for DML operations the problem of partitions locking was
resolved in commit 99fbc8e. Therefore in current work we have to complete
that fix to support partitions locking for DDL operations and other cases.
And it's in fact backport of similar commit to master 86ded36 with small
additions. Also there have been added test scenario to reproduce noted
above segfault case and related assert check inside
`calculate_table_size()` function.

Backport of PR https://github.com/greenplum-db/gpdb/pull/10468.
Fixes https://github.com/greenplum-db/gpdb/issues/12962

-------

Changes from 4ac7227 in src/backend/optimizer/prep/prepunion.c were overwritten
with changes from this commit.

(cherry-picked from commit 2763ff1)

KnightMurloc committed:
Fix use by ORCA of a newer index with HOT-chain in an older transaction. (#619)

When heap tuple is updated by legacy planner and the updated tuple is
placed at the same page (heap-only tuple, HOT), an update chain is created.
It's a chain of updated tuples, in which each tuple's ctid points to the next
tuple in the chain.

HOT chains allow to store only one index entry, which points to the first tuple
in the chain. And during Index Scan we pass through the chain, and the first
tuple visible for the current transaction is taken (for more information, see
src/backend/access/heap/README.HOT).

If we create a second index on column that has been updated, it will store the
ctid of the beginning of the existing HOT chain. If a repeatable read
transaction started before the transaction in which the second index was
created, then this index could be used in the query plan. As a result of the
search for this index, a tuple could be found that does not meet the search
condition (by a new value that is not visible to the transaction)

In the case of the legacy planner, this problem is solved the following way:

"To address this issue, regular (non-concurrent) CREATE INDEX makes the
new index usable only by new transactions and transactions that don't
have snapshots older than the CREATE INDEX command. This prevents
queries that can see the inconsistent HOT chains from trying to use the
new index and getting incorrect results. Queries that can see the index
can only see the rows that were visible after the index was created,
hence the HOT chains are consistent for them."

But ORCA does not handle this case and can use an index with a broken HOT-chain.

This patch resolves the issue for ORCA in the same way as legacy planner. During
planning we ignore newly created indexes based on their xmin.

Additionally, ORCA faced another related problem. Since ORCA has its own cache
(MD Cache) and can cache a relation object without an index that cannot be used
in the current snapshot (because MDCacheSetTransientState function returns true),
we won't be able to use the index after the problematic snapshot changes.
Therefore, we need to reset the cache after the snapshot changes in order to use
index.

This patch solves the problem in the following way: during index filtering, if
we encounter an index that we cannot use, we save TransactionXmin in the
mdcache_transaction_xmin variable. In the next queries, we check the saved xmin,
and if it is valid and differs from the current one, we reset the cache.

The create_index_hot test has also been changed. Now optimizer is turned off
before the update. Since ORCA always uses Split Update, in which case HOT chains
are not created and the problem is not reproduced. And that's why ORCA wasn't
actually tested before.

(cherry-picked from commit 5894018)

DevChattopadhyay committed:
Keep partition child lock until end of select

Introduce a new GUC gp_keep_partition_children_locks that preserves
relation locks on AO partition leaves, that are taken during planning.

This fixes concurrency issues such as #16545, where a SELECT pre-dating
a concurrent lazy vacuum can read past eof of a segfile, since the
vacuum would "drop" the segment. Adequate locking during planning is one
way to ensure that vacuumStatement_Relation() can't acquire the
AccessExclusiveLock on the coordinator. This would cause it to give up
on the drop, and the concurrent SELECT isn't compromised.

Unfortunately, we can't quite mark this GUC on by default, as it has
consequences on the amount of locks consumed:

Setting this GUC on means that we keep the acquired locks in
expand_inherited_rtentry() around longer now, up until the end of the
transaction. This can have consequences for transactions dealing with a
large number of partitions, leading us to inch closer to the
max_locks_per_transaction limit. Joins/unions etc with large partition
tables now consume more locks than before.

Notes:
(0) This conforms locking behavior to upstream behavior, that we have
in 7X (for append-optimized relations).

(1) For ORCA, we take the locks during DXL to planned statement
translation, borrowing from expand_inherited_rtentry().

(2) Unlike 7X, neither ORCA nor planner considers the statically pruned
list of partitions for locking.

(3) Currently these locks are taken only for append-optimized leaves,
but the GUC can easily be extended to cover heap tables as well, as
similar concurrency issues do exist.

Co-authored-by: Divyesh Vanjare <vanjared@vmware.com>
Co-authored-by: Soumyadeep Chakraborty <soumyadeep2007@gmail.com>
Co-authored-by: Ashwin Agrawal <aashwin@vmware.com>

-------

Missing DROP echo was added to the end of the
src/test/isolation2/expected/concurrent_select_vacuum_ao_partitioned_table.out
src/test/isolation2/expected/concurrent_select_vacuum_ao_partitioned_table_optimizer.out
files.

(cherry-picked from 4ac7227)

bandetto committed:
Revert "Fix use by ORCA of a newer index with HOT-chain in an older transaction. (#619)"

This reverts commit 5894018a785378680584751210540e94d07a32e3.

bandetto committed:
Revert "On DDL operation keep the lock on partitions until end of transaction (#295)"

This reverts commit 2763ff1446b1e0869c8ad397f3a32204b0c69dda.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67146

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67159

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1189782

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67167

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1190462

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1190463

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67276

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1195504

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1195505

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67343

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1200086

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1200087

whitehawk committed:
Add support for target cols with functional dependency on grouping ones to ORCA (#746)

Problem description: ORCA fails to handle queries with columns in the target
list, not listed in GROUP BY clause, if grouping is done by a primary key
(meaning that these columns have a functional dependency on grouping columns).
In such a case, it falls back to the standard planner. Expected behavior - ORCA
generates a plan for a query with columns in target list, if these columns have
a functional dependency on columns in GROUP BY list.

Root cause: Function CQueryMutators::ShouldFallback during query normalization
finds out that there is a target list entry, that is not a grouping column, and
it triggers fallback to the standard planner.

Fix: All columns from the target list with functional dependency on grouping
columns are added explicitly to group by clause at the query normalization stage
(at start of groupby normalization, before checking if fallback is required)
before the translation to DXL. It requires the following steps:
1) Extract such columns from target list expressions, and if they do not have a
relevant target list entry, add resjunk target list entries for them.
2) Store current unique target list entry references in groupClause - they will
be required at step 4.
3) Update all grouping sets that contain the primary key - add the functionally
dependent columns from the target list.
4) Update arguments of GROUPING functions (based on target list entry references
stored on step 2 and updated target list entry references), because they could
be shifted after step 3.

Plus, new test cases were added to the functional_deps test.

-------

An #include clause was moved in src/include/gpopt/gpdbwrappers.h to resolve
conflicts.

(cherry-picked from commit b1373f8)

bimboterminator1 commented on src/backend/optimizer/prep/prepunion.c:
```diff
@@ -1558,9 +1560,36 @@ expand_inherited_rtentry(PlannerInfo *root, RangeTblEntry *rte, Index rti)
+			 * on these 2 partition hierarchies in any way, we would be holding
+			 * (P1 + P2) locks till the end of the transaction.
+			 */
+			if (RelationIsAppendOptimized(newrelation) &&
+				(parent_is_partitioned || parent_is_interior_partition) &&
+				gp_keep_partition_children_locks)
+				releaseLockMode = NoLock;
+			else
+				releaseLockMode = rel_is_part_child(childOID) ? NoLock: lockmode;
```
Let's preserve the behaviour of DDL patch and keep holding locks as in the patch.

bimboterminator1 replied:
@bandetto, I suggest to save the code of DDL patch
https://github.com/arenadata/gpdb/blob/2763ff1446b1e0869c8ad397f3a32204b0c69dda/src/backend/optimizer/prep/prepunion.c#L1562-L1563

bimboterminator1 replied:
@bandetto, let's bring back the DDL's code. Here we'd better remove the rest of the code, and I guess I'll try to introduce this code and the dependency from GUC back in a separate patch later.

bimboterminator1 commented on src/backend/catalog/dependency.c:
```diff
@@ -1359,6 +1359,21 @@ AcquireDeletionLock(const ObjectAddress *object, int flags)
+		 * To provide at least some protection though, we still actuire
+		 * the lock momentarily.
+		 */
+		if (!gp_keep_partition_children_locks &&
```
Let's preserve the DDL's patch behaviour. Just provide correct backport of upstream patch for ORCA.

bimboterminator1 commented on src/test/isolation2/sql/ddl_on_root_locks_all_parts.sql:
```diff
@@ -22,6 +23,7 @@ CREATE TABLE ptest(i int) WITH(appendonly=true) PARTITION BY RANGE (i) (START (0
 -- session completes.
 SELECT gp_inject_fault('pg_calculate_table_size_before_pg_aoseg_estimation', 'suspend', dbid)
 FROM gp_segment_configuration WHERE role = 'p' AND content = 0;
+2: SET gp_keep_partition_children_locks TO on;
```
Here turning on the GUC value is unnecessary.

bimboterminator1 commented:
Get rid of the conflicts and fix ABI tests

bandetto committed:
Revert "Add support for target cols with functional dependency on grouping ones to ORCA (#746)"

This reverts commit b1373f861107592c4ff0792e01f60b3843d44e44.

bimboterminator1 commented on src/backend/gpopt/gpdbwrappers.cpp:
```diff
@@ -2736,47 +2739,94 @@ gpdb::IsTypeRange(Oid typid)
 }
 
-RowMarkClause *
-gpdb::GetParseRowmark(Query *query, Index rtindex)
```
We could resolve conflicts without removing the previous code and just change its location inside a file. I guess it is legit to make such changes for **our** patches.  We **do not need** a separate commit 2809cad07593997680149fdffca9fb3f1c89b2b8.

bimboterminator1 commented on src/test/isolation2/sql/ddl_on_root_locks_all_parts.sql:
```diff
@@ -11,6 +11,7 @@ CREATE TABLE ptest(i int) WITH(appendonly=true) PARTITION BY RANGE (i) (START (0
 -- The first session removes partitioned table inside transaction block
 -- When locks on partitions are not set (buggy case), the second session that
 -- works with specific partition will have progress
+1: SET gp_keep_partition_children_locks TO on;
```
Do that while applying DDL patch. We do not need this commit

bimboterminator1 replied:
We don't need to execute this command. We can remove it. And look at the comment about the prepunion code

bimboterminator1 commented on src/test/isolation2/expected/concurrent_select_vacuum_ao_partitioned_table.out:
```diff
@@ -233,3 +233,4 @@ VACUUM
  1 | 1 | 2 
 (1 row)
 DROP TABLE foo_part;
+DROP
```
Unfortunately, if the tests do not pass without DROP, we have to add DROP to the **upstream patch**. I don't know why the tests have passed in upstream, but in this PR each separate commit should bring the whole project to the valid state. So, we have no choice but fixing the tests inside the upstream commit.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/68109

BenderArenadata commented:
Failed job ORCA code linter on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1231846

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1231843

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1231844

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1231834

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/68122

BenderArenadata commented:
Failed job ORCA code linter on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1233712

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1233710

bimboterminator1 commented:
Can you fix clang-format tests? Probably inside the whitehawk's patch

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/68161

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1233899

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1233900

RekGRpth commented:
I think every backport should contain a link/hash of the original commit, as well as a description of what was changed from the original commit and why.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/68252

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1237188

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1237189

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-4950

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/68448

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1246200

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1246201

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1246198
